pkg/controller/core/cohort_controller.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
+	config "sigs.k8s.io/kueue/apis/config/v1beta1"
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/queue"
+)
+
+// CohortReconciler is responsible for synchronizing the in-memory
+// representation of Cohorts in cache.Cache and queue.Manager with
+// Cohort Kubernetes objects.
+type CohortReconciler struct {
+	client   client.Client
+	log      logr.Logger
+	cache    *cache.Cache
+	qManager *queue.Manager
+}
+
+func NewCohortReconciler(client client.Client, cache *cache.Cache, qManager *queue.Manager) CohortReconciler {
+	return CohortReconciler{client, ctrl.Log.WithName("cohort-reconciler"), cache, qManager}
+}
+
+func (r *CohortReconciler) SetupWithManager(mgr ctrl.Manager, cfg *config.Configuration) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&kueue.Cohort{}).
+		WithOptions(controller.Options{NeedLeaderElection: ptr.To(false)}).
+		WithEventFilter(r).
+		Complete(WithLeadingManager(mgr, r, &kueue.Cohort{}, cfg))
+}
+
+func (r *CohortReconciler) Create(e event.CreateEvent) bool {
+	return true
+}
+
+func (r *CohortReconciler) Update(e event.UpdateEvent) bool {
+	oldCohort, ok := e.ObjectOld.(*kueue.Cohort)
+	if !ok {
+		return false
+	}
+	newCohort, ok := e.ObjectNew.(*kueue.Cohort)
+	if !ok {
+		return false
+	}
+	log := r.log.WithValues("cohort", klog.KObj(newCohort))
+	if equality.Semantic.DeepEqual(oldCohort.Spec.ResourceGroups, newCohort.Spec.ResourceGroups) &&
+		oldCohort.Spec.Parent == newCohort.Spec.Parent {
+		log.V(2).Info("Filtering Cohort update event as Cohort unchanged")
+		return false
+	}
+	log.V(2).Info("Processing Cohort update event")
+	return true
+}
+
+func (r *CohortReconciler) Delete(e event.DeleteEvent) bool {
+	return true
+}
+
+func (r *CohortReconciler) Generic(e event.GenericEvent) bool {
+	return true
+}
+
+//+kubebuilder:rbac:groups=kueue.x-k8s.io,resources=cohorts,verbs=get;list;watch;create;update;patch;delete
+
+func (r *CohortReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
Because you are doing all operations in the reconciler, competing with status updates, which can take a relative long time, we might need a higher number of workers to reduce the delay between a change and them applying to the admission decisions.

Reply:
add a reasonable number to the reconcilers configuration https://github.com/kubernetes-sigs/kueue/blob/d4331979e09c7096f1f17426c9551476b3fe77fc/config/components/manager/controller_manager_config.yaml#L14

Reply:
Set to 1, matching ClusterQueue. Users expecting many Cohorts can configure this number to be greater.